File: backends/spring-java/src/test/java/dev/stackverse/backend/message/MessageLocalizerTest.java
package dev.stackverse.backend.message;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import jakarta.servlet.http.HttpServletRequest;
import java.time.Instant;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpHeaders;

class MessageLocalizerTest {
    private final MessageRepository repository = mock(MessageRepository.class);
    private final LanguageResolver languageResolver = mock(LanguageResolver.class);
    private final MessageLocalizer localizer = new MessageLocalizer(repository, languageResolver);
    private final HttpServletRequest request = mock(HttpServletRequest.class);

    @Test
    void returnsMessageInResolvedLanguage() {
        when(request.getParameter("lang")).thenReturn("pl");
        when(request.getHeader(HttpHeaders.ACCEPT_LANGUAGE)).thenReturn("en;q=0.8");
        when(languageResolver.resolve("pl", "en;q=0.8")).thenReturn("pl");
        when(repository.findByKeyAndLanguage("validation.title.required", "pl"))
            .thenReturn(message("validation.title.required", "pl", "Podaj tytul."));

        assertThat(localizer.localize("validation.title.required", request)).isEqualTo("Podaj tytul.");

        verify(languageResolver).resolve("pl", "en;q=0.8");
    }

    @Test
    void fallsBackToEnglishWhenResolvedLanguageLacksKey() {
        when(languageResolver.resolve(null, null)).thenReturn("pl");
        when(repository.findByKeyAndLanguage("validation.url.invalid", "pl")).thenReturn(null);
        when(repository.findByKeyAndLanguage("validation.url.invalid", "en"))
            .thenReturn(message("validation.url.invalid", "en", "Enter a valid URL."));

        assertThat(localizer.localize("validation.url.invalid", request)).isEqualTo("Enter a valid URL.");
    }

    @Test
    void fallsBackToKeyWhenNoMessageExists() {
        when(languageResolver.resolve(null, null)).thenReturn("en");
        when(repository.findByKeyAndLanguage("validation.missing", "en")).thenReturn(null);

        assertThat(localizer.localize("validation.missing", request)).isEqualTo("validation.missing");
    }

    private static Message message(String key, String language, String text) {
        return new Message(key, language, text, null, Instant.EPOCH, Instant.EPOCH);
    }
}

File: backends/spring-kotlin/src/test/kotlin/dev/stackverse/backend/message/MessageLocalizerTest.kt
package dev.stackverse.backend.message

import jakarta.servlet.http.HttpServletRequest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import org.mockito.Mockito.`when`
import java.time.Instant

class MessageLocalizerTest {

    private val repository = mock(MessageRepository::class.java)
    private val languageResolver = mock(LanguageResolver::class.java)
    private val request = mock(HttpServletRequest::class.java)
    private val localizer = MessageLocalizer(repository, languageResolver)

    @Test
    fun `uses the text for the language resolved from the request`() {
        `when`(request.getParameter("lang")).thenReturn("pl")
        `when`(request.getHeader("Accept-Language")).thenReturn("en;q=0.8")
        `when`(languageResolver.resolve("pl", "en;q=0.8")).thenReturn("pl")
        `when`(repository.findByKeyAndLanguage("validation.title.required", "pl"))
            .thenReturn(message("pl", "Tytuł jest wymagany."))

        assertThat(localizer.localize("validation.title.required", request)).isEqualTo("Tytuł jest wymagany.")

        verify(repository).findByKeyAndLanguage("validation.title.required", "pl")
        verifyNoMoreInteractions(repository)
    }

    @Test
    fun `falls back to english when the resolved language lacks the key`() {
        `when`(languageResolver.resolve(null, "pl")).thenReturn("pl")
        `when`(request.getHeader("Accept-Language")).thenReturn("pl")
        `when`(repository.findByKeyAndLanguage("validation.url.invalid", "pl")).thenReturn(null)
        `when`(repository.findByKeyAndLanguage("validation.url.invalid", DEFAULT_LANGUAGE))
            .thenReturn(message(DEFAULT_LANGUAGE, "Enter a valid URL."))

        assertThat(localizer.localize("validation.url.invalid", request)).isEqualTo("Enter a valid URL.")
    }

    @Test
    fun `falls back to the stable key when no localized text exists`() {
        `when`(languageResolver.resolve(null, null)).thenReturn("pl")
        `when`(repository.findByKeyAndLanguage("validation.unknown", "pl")).thenReturn(null)
        `when`(repository.findByKeyAndLanguage("validation.unknown", DEFAULT_LANGUAGE)).thenReturn(null)

        assertThat(localizer.localize("validation.unknown", request)).isEqualTo("validation.unknown")

        verify(repository).findByKeyAndLanguage("validation.unknown", "pl")
        verify(repository).findByKeyAndLanguage("validation.unknown", DEFAULT_LANGUAGE)
        verifyNoMoreInteractions(repository)
    }

    private fun message(language: String, text: String) = Message(
        key = "validation.test",
        language = language,
        text = text,
        description = null,
        createdAt = Instant.EPOCH,
        updatedAt = Instant.EPOCH,
    )
}
